### PR TITLE
MGMT-9470: Publish catalog index to quay

### PR DESCRIPTION
**Pull request by adriengentil**
The current solution build a bundle that points to the latest tag of the assisted-service images which makes its content unpredictable. This new job pins the assisted images in the bundle and publishes a catalog index created from it.

In order to work around the docker-in-docker situation, the job uses `umoci` along with `skopeo` to build and publish the bundle and the catalog images.

### CODE REVIEW

**Comment by omertuc** [comment]
So can we merge this?

**Comment by omertuc** [comment]
/retitle MGMT-9470: Publish catalog index to quay

File: ci-operator/step-registry/assisted/baremetal/operator/catalog-publish/assisted-baremetal-operator-catalog-publish-commands.sh [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+/tmp/umoci config --config.label 'operators.operatorframework.io.bundle.mediatype.v1=registry+v1' --image bundle:latest
+/tmp/umoci config --config.label 'operators.operatorframework.io.bundle.manifests.v1=manifests/' --image bundle:latest
+/tmp/umoci config --config.label 'operators.operatorframework.io.bundle.metadata.v1=metadata/' --image bundle:latest
+/tmp/umoci config --config.label 'operators.operatorframework.io.bundle.package.v1=assisted-service-operator' --image bundle:latest
+/tmp/umoci config --config.label 'operators.operatorframework.io.bundle.channels.v1=alpha,ocm-2.5' --image bundle:latest
+/tmp/umoci config --config.label 'operators.operatorframework.io.bundle.channel.default.v1=alpha' --image bundle:latest
+/tmp/umoci config --config.label 'operators.operatorframework.io.metrics.builder=operator-sdk-v1.10.0+git' --image bundle:latest
+/tmp/umoci config --config.label 'operators.operatorframework.io.metrics.mediatype.v1=metrics+v1' --image bundle:latest
+/tmp/umoci config --config.label 'operators.operatorframework.io.metrics.project_layout=go.kubebuilder.io/v3' --image bundle:latest
+/tmp/umoci unpack --rootless --image bundle:latest bundle
```
Comment (omertuc) [inline ci-operator/step-registry/assisted/baremetal/operator/catalog-publish/assisted-baremetal-operator-catalog-publish-commands.sh:51]: Ideally we should probably put this inside script files in the service repo, but for now let's merge this

File: ci-operator/step-registry/assisted/baremetal/operator/catalog-publish/assisted-baremetal-operator-catalog-publish-ref.yaml [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+  resources:
+    requests:
+      cpu: 1000m
+      memory: 500Mi
+  credentials:
+  - mount_path: /secrets/docker
+    name: edge-infrastructure-quay-push
+    namespace: test-credentials
+  - mount_path: /tmp/secret
+    name: agentil-quay-creds
```
Comment (osherdp) [inline ci-operator/step-registry/assisted/baremetal/operator/catalog-publish/assisted-baremetal-operator-catalog-publish-ref.yaml:15]: we need this in addition to ``edge-infrastructure-quay-push``?

File: ci-operator/step-registry/assisted/baremetal/operator/catalog-publish/assisted-baremetal-operator-catalog-publish-commands.sh [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+/tmp/umoci config --config.label 'operators.operatorframework.io.bundle.mediatype.v1=registry+v1' --image bundle:latest
+/tmp/umoci config --config.label 'operators.operatorframework.io.bundle.manifests.v1=manifests/' --image bundle:latest
+/tmp/umoci config --config.label 'operators.operatorframework.io.bundle.metadata.v1=metadata/' --image bundle:latest
+/tmp/umoci config --config.label 'operators.operatorframework.io.bundle.package.v1=assisted-service-operator' --image bundle:latest
+/tmp/umoci config --config.label 'operators.operatorframework.io.bundle.channels.v1=alpha,ocm-2.5' --image bundle:latest
+/tmp/umoci config --config.label 'operators.operatorframework.io.bundle.channel.default.v1=alpha' --image bundle:latest
+/tmp/umoci config --config.label 'operators.operatorframework.io.metrics.builder=operator-sdk-v1.10.0+git' --image bundle:latest
+/tmp/umoci config --config.label 'operators.operatorframework.io.metrics.mediatype.v1=metrics+v1' --image bundle:latest
+/tmp/umoci config --config.label 'operators.operatorframework.io.metrics.project_layout=go.kubebuilder.io/v3' --image bundle:latest
```
Comment (osherdp) [inline ci-operator/step-registry/assisted/baremetal/operator/catalog-publish/assisted-baremetal-operator-catalog-publish-commands.sh:50]: not so sure what these mean, maybe just a simple comment can help here

File: ci-operator/step-registry/assisted/baremetal/operator/catalog-publish/assisted-baremetal-operator-catalog-publish-commands.sh [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+
+/tmp/opm index add \
+    --bundles "${BUNDLE_DESTINATION_DIGEST}" \
+    --out-dockerfile index.Dockerfile \
+    --generate
+
+skopeo copy "docker://${OPM_BASE_IMAGE}" oci:opm:latest
+/tmp/umoci unpack --rootless --image opm:latest bundle
+mv database bundle/rootfs/
+/tmp/umoci repack --image opm:index bundle
```
Comment (osherdp) [inline ci-operator/step-registry/assisted/baremetal/operator/catalog-publish/assisted-baremetal-operator-catalog-publish-commands.sh:78]: optional: you can set ``export PATH="${PATH}":/tmp/bin`` and have umochi/opm in there (to improve appearance)

**Review by osherdp — CHANGES_REQUESTED** [review]

File: ci-operator/step-registry/assisted/baremetal/operator/catalog-publish/assisted-baremetal-operator-catalog-publish-ref.yaml (hunk 2 above)
Comment (adriengentil) [reply ci-operator/step-registry/assisted/baremetal/operator/catalog-publish/assisted-baremetal-operator-catalog-publish-ref.yaml:15]: nope :fearful:

File: ci-operator/step-registry/assisted/baremetal/operator/catalog-publish/assisted-baremetal-operator-catalog-publish-commands.sh (hunk 4 above)
Comment (omertuc) [reply ci-operator/step-registry/assisted/baremetal/operator/catalog-publish/assisted-baremetal-operator-catalog-publish-commands.sh:78]: The other way around `export PATH="/tmp/bin:${PATH}"` if we want to ensure we're using the one we downloaded for ourselves

File: ci-operator/step-registry/assisted/baremetal/operator/catalog-publish/assisted-baremetal-operator-catalog-publish-commands.sh (hunk 3 above)
Comment (omertuc) [reply ci-operator/step-registry/assisted/baremetal/operator/catalog-publish/assisted-baremetal-operator-catalog-publish-commands.sh:50]: A link saying this is basically a umoci recreation of this dockerfile https://github.com/openshift/assisted-service/blob/f5b56b1923b534c16a13c17ef29bde902466e4cc/deploy/olm-catalog/bundle.Dockerfile would probably be best

Comment (adriengentil) [reply ci-operator/step-registry/assisted/baremetal/operator/catalog-publish/assisted-baremetal-operator-catalog-publish-commands.sh]: I added few comments and added a reference to the dockerfile of the bundle

File: ci-operator/step-registry/assisted/baremetal/operator/catalog-publish/assisted-baremetal-operator-catalog-publish-ref.yaml (hunk 2 above)
Comment (adriengentil) [reply ci-operator/step-registry/assisted/baremetal/operator/catalog-publish/assisted-baremetal-operator-catalog-publish-ref.yaml]: I removed it

File: ci-operator/step-registry/assisted/baremetal/operator/catalog-publish/assisted-baremetal-operator-catalog-publish-commands.sh [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+cd "${tmp}"
+
+# Create the index database from the bundle we just published
+opm index add \
+    --bundles "${BUNDLE_DESTINATION_DIGEST}" \
+    --out-dockerfile index.Dockerfile \
+    --generate
+
+# Add the index database to the opm base image
+# It does the same operation as in the Dockerfile generated by the opm index add... command above
```
Comment (omertuc) [inline ci-operator/step-registry/assisted/baremetal/operator/catalog-publish/assisted-baremetal-operator-catalog-publish-commands.sh:85]: This comment should be lower down, I believe

File: ci-operator/step-registry/assisted/baremetal/operator/catalog-publish/assisted-baremetal-operator-catalog-publish-commands.sh (hunk 4 above)
Comment (adriengentil) [reply ci-operator/step-registry/assisted/baremetal/operator/catalog-publish/assisted-baremetal-operator-catalog-publish-commands.sh]: Done

File: ci-operator/step-registry/assisted/baremetal/operator/catalog-publish/assisted-baremetal-operator-catalog-publish-commands.sh [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+skopeo copy "docker://${OPM_BASE_IMAGE}" oci:opm:latest
+umoci unpack --rootless --image opm:latest bundle
+mv database bundle/rootfs/
+umoci repack --image opm:index bundle
+
+# Configure the index image and publish it
```
Comment (omertuc) [inline ci-operator/step-registry/assisted/baremetal/operator/catalog-publish/assisted-baremetal-operator-catalog-publish-commands.sh:91]:
```suggestion
```

Comment (omertuc) [reply ci-operator/step-registry/assisted/baremetal/operator/catalog-publish/assisted-baremetal-operator-catalog-publish-commands.sh:91]: Just to make sure the comment above looks like it applies to the lines below as well

Comment (adriengentil) [reply ci-operator/step-registry/assisted/baremetal/operator/catalog-publish/assisted-baremetal-operator-catalog-publish-commands.sh]: Like that?

Comment (omertuc) [reply ci-operator/step-registry/assisted/baremetal/operator/catalog-publish/assisted-baremetal-operator-catalog-publish-commands.sh]: yep

**Review by osherdp — APPROVED** [review]
@adriengentil one thing that we'll need to handle (not now, it can be a followup) is to make it not rehearse (or have no actual image promotion if it rehearses)

**Comment by adriengentil** [comment]
> @adriengentil one thing that we'll need to handle (not now, it can be a followup) is to make it not rehearse (or have no actual image promotion if it rehearses)

@osherdp postsubmit jobs are configured to not be rehearsable by default: https://docs.ci.openshift.org/docs/architecture/ci-operator/#post-submit-tests

**Comment by osherdp** [comment]
> > @adriengentil one thing that we'll need to handle (not now, it can be a followup) is to make it not rehearse (or have no actual image promotion if it rehearses)
> 
> @osherdp postsubmit jobs are configured to not be rehearsable by default: https://docs.ci.openshift.org/docs/architecture/ci-operator/#post-submit-tests

That's great
Thanks for checking that up

**Pull request merged**